datumaro/plugins/coco_format/extractor.py

```diff
@@ -54,145 +59,134 @@ def __init__(self, path, task, *,
 
         self._merge_instance_polygons = merge_instance_polygons
 
+        json_data = self._load_json(path)
+        self._label_map = {} # coco_id -> dm_id
+        self._load_categories(json_data,
+            keep_original_ids=keep_original_category_ids,
+        )
+
         if self._task == CocoTask.panoptic:
             #panoptic is not added to pycocotools
-            panoptic_config = self._load_panoptic_config(path)
-            panoptic_images = osp.splitext(path)[0]
-
-            self._load_label_categories(
-                panoptic_config['categories'],
-                keep_original_ids=keep_original_category_ids,
-            )
-
-            self._items = list(self._load_panoptic_items(panoptic_config,
-                panoptic_images).values())
+            self._items = self._load_panoptic_items(json_data,
+                mask_dir=osp.splitext(path)[0])
         else:
-            loader = self._make_subset_loader(path)
-            self._load_categories(
-                loader,
-                keep_original_ids=keep_original_category_ids,
-            )
-            self._items = list(self._load_items(loader).values())
-
-    @staticmethod
-    def _make_subset_loader(path):
-        # TODO: remove pycocotools dependency?
+            self._items = self._load_items(json_data)
```

Comment:
Is there a reason to no longer convert `self._items` to a list? I don't see its keys being used anywhere after `_load_items` completes.

Reply:
I'm not sure I got the question, but there is no point in converting it to list.

Reply:
The point is to free up the memory used for the keys and for the dictionary internals.

I would also say that changing `_items` to a `dict` breaks the contract established by `SourceExtractor`. It happens to work now, but it could blow up if `SourceExtractor` is modified.

Reply:
> The point is to free up the memory used for the keys and for the dictionary internals.

This optimization is not effective.

> I would also say that changing _items to a dict breaks the contract established by SourceExtractor. It happens to work > now, but it could blow up if SourceExtractor is modified.

It's possible, but it's unlikely to happen occasionally and unnoticed.